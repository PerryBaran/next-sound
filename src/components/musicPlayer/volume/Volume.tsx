"use client"

import Image from "next/image"
import css from "./volume.module.css"
import { volumeLow, volumeMedium, volumeMute } from "../../../media/icons"
import { useState } from "react"

interface Props {
  volume: number
  handleVolume: (value: number) => void
}

export default function Volume({ volume, handleVolume }: Props) {
  const [toggleVolume, setToggleVolume] = useState(volume)

  const changeVolume = (e: React.ChangeEvent<HTMLInputElement>) => {
    const currentVolume = Number(e.target.value) / 100
    handleVolume(currentVolume)
  }

  const handleVolumeToggle = (
    e: any
  ) => {
    if (e.nativeEvent?.pointerType !== "mouse") return
    const preVolume = volume
    if (preVolume !== 0) {
      handleVolume(0)
      setToggleVolume(preVolume)
    } else {
      handleVolume(toggleVolume)
    }
  }

  return (
    <div className={css["container"]}>
      <input
        type="range"
        name="volume"
        min={0}
        max={100}
        value={volume * 100}
        onChange={changeVolume}
      />
      <div className={css["volume"]}>
        <button type="button" onClick={handleVolumeToggle}>
          <VolumeIcon volume={volume} />
        </button>
      </div>
    </div>
  )
}

function VolumeIcon({ volume }: { volume: number }) {
  if (volume === 0)
    return <Image src={volumeMute} alt="volume muted" height={30} width={30} />

  if (volume < 0.5)
    return <Image src={volumeLow} alt="low volume" height={30} width={30} />

  return <Image src={volumeMedium} alt="high volume" height={30} width={30} />
}
