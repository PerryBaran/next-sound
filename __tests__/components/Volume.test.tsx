import { fireEvent, render, screen } from "@testing-library/react"
import { useState } from "react"
import Volume from "../../src/components/musicPlayer/volume/Volume"

describe("Volume", () => {
  let props = {
    volume: 0.5,
    handleVolume: jest.fn()
  }

  beforeEach(() => {
    props.handleVolume = jest.fn()
  })

  test("renders button", () => {
    render(<Volume {...props} />)

    expect(screen.getByRole("button")).toBeInTheDocument()
  })

  test("renders correctly when volume is equal to 0", () => {
    const volume = 0
    render(<Volume {...props} volume={volume} />)

    expect(screen.getByAltText("volume muted")).toBeInTheDocument()
    expect(screen.getByDisplayValue(volume * 100)).toBeInTheDocument()
  })

  test("renders correctly when volume is less than 0.5 but greater than 0", () => {
    const volume = 0.3
    render(<Volume {...props} volume={volume} />)

    expect(screen.getByAltText("low volume")).toBeInTheDocument()
    expect(screen.getByDisplayValue(volume * 100)).toBeInTheDocument()
  })

  test("renders correctly when volume is greater than or equal to 0.5", () => {
    const volume = 0.5
    render(<Volume {...props} volume={volume} />)

    expect(screen.getByAltText("high volume")).toBeInTheDocument()
    expect(screen.getByDisplayValue(volume * 100)).toBeInTheDocument()
  })

  test("button click", () => {
    const TestVolumeToggle = ({
      volume,
      handleVolume
    }: {
      volume: number
      handleVolume: (value: number) => void
    }) => {
      const [mockVolume, setMockVolume] = useState(volume)

      const mockHandleVolume = (value: number) => {
        handleVolume(value)
        setMockVolume(value)
      }

      return <Volume volume={mockVolume} handleVolume={mockHandleVolume} />
    }
    render(<TestVolumeToggle {...props} />)
    const button = screen.getByRole("button")

    expect(props.handleVolume).toBeCalledTimes(0)

    fireEvent.click(button)

    expect(props.handleVolume).toBeCalledTimes(1)
    expect(props.handleVolume).toHaveBeenCalledWith(0)

    fireEvent.click(button)

    expect(props.handleVolume).toBeCalledTimes(2)
    expect(props.handleVolume).toHaveBeenCalledWith(props.volume)
  })

  test("input change", () => {
    render(<Volume {...props} />)

    expect(props.handleVolume).toBeCalledTimes(0)

    const newValue = 20
    fireEvent.change(screen.getByDisplayValue(props.volume * 100), {
      target: { value: newValue }
    })

    expect(props.handleVolume).toBeCalledTimes(1)
    expect(props.handleVolume).toHaveBeenCalledWith(newValue / 100)
  })

  test("snapshot", () => {
    const { asFragment } = render(<Volume {...props} />)

    expect(asFragment()).toMatchSnapshot()
  })
})
