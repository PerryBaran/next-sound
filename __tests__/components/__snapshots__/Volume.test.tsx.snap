// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Volume snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <input
      max="100"
      min="0"
      name="volume"
      type="range"
      value="50"
    />
    <div
      class="volume"
    >
      <button
        type="button"
      >
        <img
          alt="high volume"
          data-nimg="1"
          decoding="async"
          height="30"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=64&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=64&q=75 2x"
          style="color: transparent;"
          width="30"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
